repository/Seaside-WebSocket-Examples.package/WAWebSocketExampleBroadcaster.class/instance/handleWebSocketSetup.in..st

```diff
@@ -0,0 +1,19 @@
+handling
+handleWebSocketSetup: aWAWebSocket in: aRequestContext
+	
+	"webSocket timeout: 300." "5 minutes timeout"
+	
+	Transcript cr; show: 'New ', aWAWebSocket printString.
+
+	aWAWebSocket onMessage:[ :data |
+		self socketsDo: [ :socket | socket send: data ] ].
```

Comment:
I'm a bit concerned about the socket iteration pattern.

- In this case, and probably others, since `#socketsDo:` acquires a mutex it blocks until data has been sent to all sockets. I wonder if it would be beneficial to have some sort of job/task queue and submitting a task for every socket and hold the mutex only for submitting the tasks.
- Between when we decided to send data to the sockets and actually doing it the socket may have become closed or invalid which makes we wonder whether we should:
  - check each web socket for validity in `#send:` and possibly removing it
  - wrap each `#send:` in an `#on:do:`
- Having all the sockets in an array, linearly scanning over them and hoping `#onClose:` will always get sent and the web socket removed. I'm concerned with scalability and that we end up with dead or closed web sockets in the array that never get removed.

Reply:
yes, all valid concerns and this still needs to be addressed. In addition, the seaside session needs to be kept alive as long as the sockets are communicating/existing. While doing experiments with the simple counter example, I encounter websocket connections being closed without any callback firing without a consistent timeout or other immediately visible reason. So, I still have a long way to go to make sure this works in practice.